### PR TITLE
Refactored handling/updating of the ActivityRegistry

### PR DESCRIPTION
**Pull request by raymonddenhaan**
Refactored handling/updating of the ActivityRegistry
- Complete refresh of the registry is done by building a new dictionary, then swapping it. Instead of clearing the working version.
- Creation/Publishing/Retracting/Deleting of a workflow only leads to a change to that specific workflow in the registry instead of recreating the registry from scratch
- Added a failover that on a mis in the registry the entire registry is going to be refreshed.

- - -
This change is [Reviewable](https://reviewable.io/reviews/elsa-workflows/elsa-core/5468)

### CODE REVIEW

**Comment by raymonddenhaan** [comment]
closes #5464

File: src/modules/Elsa.Dsl/Contracts/IFunctionActivityRegistry.cs [hunk 1]
```diff
@@ -22,12 +22,12 @@ public interface IFunctionActivityRegistry
...
-    
+
     /// <summary>
     /// Resolves a function to an activity that can be invoked from a DSL script.
     /// </summary>
     /// <param name="functionName">The name of the function.</param>
     /// <param name="arguments">The arguments that are passed to the function.</param>
     /// <returns>An activity that can be invoked from a DSL script.</returns>
-    IActivity ResolveFunction(string functionName, IEnumerable<object?>? arguments = default);
+    Task<IActivity> ResolveFunction(string functionName, IEnumerable<object?>? arguments = default);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Dsl/Contracts/IFunctionActivityRegistry.cs:32]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Dsl/Interpreters/WorkflowDefinitionBuilderInterpreter/VisitObject.cs [hunk 2]
```diff
@@ -38,12 +38,12 @@ public override IWorkflowBuilder VisitObject(ElsaParser.ObjectContext context)
         return DefaultResult;
     }
 
-    private object GetObject(ElsaParser.ObjectContext context)
+    private async Task<object> GetObject(ElsaParser.ObjectContext context)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Dsl/Interpreters/WorkflowDefinitionBuilderInterpreter/VisitObject.cs:41]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.JavaScript/Services/JintJavaScriptEvaluator.cs [hunk 3]
```diff
@@ -171,11 +171,11 @@ private static void CreateVariableAccessors(Engine engine, ExpressionExecutionCo
         }
     }
 
-    private static void CreateActivityOutputAccessors(Engine engine, ExpressionExecutionContext context)
+    private static async Task CreateActivityOutputAccessors(Engine engine, ExpressionExecutionContext context)
```
Comment (sfmskywalker) [inline src/modules/Elsa.JavaScript/Services/JintJavaScriptEvaluator.cs:174]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Contexts/WorkflowExecutionContext.cs [hunk 4]
```diff
@@ -572,9 +579,9 @@ internal void TransitionTo(WorkflowSubStatus subStatus)
     /// <summary>
     /// Creates a new <see cref="ActivityExecutionContext"/> for the specified activity.
     /// </summary>
-    public ActivityExecutionContext CreateActivityExecutionContext(IActivity activity, ActivityInvocationOptions? options = default)
+    public async Task<ActivityExecutionContext> CreateActivityExecutionContext(IActivity activity, ActivityInvocationOptions? options = default)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Contexts/WorkflowExecutionContext.cs:582]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Contracts/IActivityRegistryLookupService.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+/// Represents a service for looking up activity descriptors.
+/// </summary>
+public interface IActivityRegistryLookupService
+{
+    /// <summary>
+    /// Returns the activity descriptor for the specified activity type.
+    /// </summary>
+    /// <param name="type">The activity type.</param>
+    /// <returns>The activity descriptor for the specified activity type.</returns>
+    Task<ActivityDescriptor?> Find(string type);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Contracts/IActivityRegistryLookupService.cs:15]: Please mind the naming convention to apply the `Async` suffix when declaring async methods.

File: src/modules/Elsa.Workflows.Core/Contracts/IIdentityGraphService.cs [hunk 6]
```diff
@@ -21,17 +21,17 @@ public interface IIdentityGraphService
     /// <summary>
     /// Assign unique identities tot the specified <see cref="ActivityNode"/> graph.
     /// </summary>
-    void AssignIdentities(ActivityNode root);
+    Task AssignIdentities(ActivityNode root);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Contracts/IIdentityGraphService.cs:24]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Contracts/IWorkflowStateExtractor.cs [hunk 7]
```diff
@@ -13,12 +13,12 @@ public interface IWorkflowStateExtractor
...
-    
+
     /// <summary>
     /// Applies workflow state to a workflow execution context.
     /// </summary>
     /// <param name="workflowExecutionContext">The workflow execution context to map onto.</param>
     /// <param name="state">The workflow state to apply.</param>
     /// <returns>The updated workflow execution context.</returns>
-    WorkflowExecutionContext Apply(WorkflowExecutionContext workflowExecutionContext, WorkflowState state);
+    Task<WorkflowExecutionContext> Apply(WorkflowExecutionContext workflowExecutionContext, WorkflowState state);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Contracts/IWorkflowStateExtractor.cs:23]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistryLookupService.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        return Find(() => activityRegistry.Find(predicate));
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ActivityDescriptor> FindMany(Func<ActivityDescriptor, bool> predicate)
+    {
+        return activityRegistry.FindMany(predicate);
+    }
+    
+    private async Task<ActivityDescriptor?> Find(Func<ActivityDescriptor?> findPredicate)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Services/ActivityRegistryLookupService.cs:35]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Services/WorkflowStateExtractor.cs [hunk 9]
```diff
@@ -96,10 +96,10 @@ private void ApplyProperties(WorkflowState state, WorkflowExecutionContext workf
             workflowExecutionContext.Properties[property.Key] = property.Value;
     }
 
-    private static void ApplyActivityExecutionContexts(WorkflowState state, WorkflowExecutionContext workflowExecutionContext)
+    private static async Task ApplyActivityExecutionContexts(WorkflowState state, WorkflowExecutionContext workflowExecutionContext)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Services/WorkflowStateExtractor.cs:99]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Core/Services/WorkflowStateExtractor.cs [hunk 10]
```diff
@@ -128,7 +128,7 @@ private static void ApplyActivityExecutionContexts(WorkflowState state, Workflow
         workflowExecutionContext.ActivityExecutionContexts = activityExecutionContexts;
         return;
 
-        ActivityExecutionContext? CreateActivityExecutionContext(ActivityExecutionContextState activityExecutionContextState)
+        async Task<ActivityExecutionContext?> CreateActivityExecutionContext(ActivityExecutionContextState activityExecutionContextState)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Services/WorkflowStateExtractor.cs:131]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Management/Activities/WorkflowDefinitionActivity/WorkflowDefinitionActivity.cs [hunk 11]
```diff
@@ -84,10 +84,10 @@ private async ValueTask OnChildCompletedAsync(ActivityCompletedContext context)
         await activityExecutionContext.CompleteActivityAsync(completeCompositeSignal?.Value);
     }
 
-    private void CopyInputOutputToVariables(ActivityExecutionContext context)
+    private async Task CopyInputOutputToVariables(ActivityExecutionContext context)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Activities/WorkflowDefinitionActivity/WorkflowDefinitionActivity.cs:87]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Management/Activities/WorkflowDefinitionActivity/WorkflowDefinitionActivity.cs [hunk 12]
```diff
@@ -165,10 +165,10 @@ await workflowDefinitionService.FindWorkflowGraphAsync(filter, cancellationToken
         return workflowGraph;
     }
 
-    private ActivityDescriptor FindActivityDescriptor(IServiceProvider serviceProvider)
+    private async Task<ActivityDescriptor> FindActivityDescriptor(IServiceProvider serviceProvider)
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Activities/WorkflowDefinitionActivity/WorkflowDefinitionActivity.cs:168]: Please mind the naming convention to apply the `Async` suffix when changing methods into async.

File: src/modules/Elsa.Workflows.Management/Contracts/IActivityRegistryPopulator.cs [hunk 13]
```diff
@@ -27,4 +20,21 @@ public interface IActivityRegistryPopulator
...
+    void RemoveDefinitionFromRegistry(Type providerType, string workflowDefinitionId, CancellationToken cancellationToken = default);
+    
+    
+    /// <summary>
+    /// Removes a workflow definition version activity from the <see cref="IActivityRegistry"/>.
+    /// </summary>
+    /// <param name="providerType">The type of the Activity Provider.</param>
+    /// <param name="workflowDefinitionVersionId">The ID of the workflow definition to remove.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    void RemoveDefinitionVersionFromRegistry(Type providerType, string workflowDefinitionVersionId, CancellationToken cancellationToken = default);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Contracts/IActivityRegistryPopulator.cs:39]: The IActivityRegistry shouldn't be aware of workflow definitions; it should only be aware of activity descriptor providers.

I missed this the first time the first version of `AddToRegistry(..., workflowdefinitionId)` was added on April 10th, which is my bad. I think this is a mistake and should be revisited.

File: src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdate.cs [hunk 14]
```diff
@@ -0,0 +1,6 @@
+namespace Elsa.Workflows.Management.Notifications;
+
+/// <summary>
+/// Represents an update to a specific workflow definition version.
+/// </summary>
+public record WorkflowDefinitionVersionsUpdate(string Id, string DefinitionId, bool UsableAsActivity);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdate.cs:6]: Given that this type represents an update to a single workflow definition version, the name should changed to `WorkflowDefinitionVersionUpdate`.

File: src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdated.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+using Elsa.Mediator.Contracts;
+using JetBrains.Annotations;
+
+namespace Elsa.Workflows.Management.Notifications;
+
+/// <summary>
+/// A notification that is sent when specific workflow definition versions are updated.
+/// </summary>
+[PublicAPI]
+public record WorkflowDefinitionVersionsUpdated(IEnumerable<WorkflowDefinitionVersionsUpdate> VersionUpdate) : INotification;
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdated.cs:10]: `VersionUpdate` should be renamed to `VersionUpdates` (plural).

File: src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdating.cs [hunk 16]
```diff
@@ -0,0 +1,10 @@
+using Elsa.Mediator.Contracts;
+using JetBrains.Annotations;
+
+namespace Elsa.Workflows.Management.Notifications;
+
+/// <summary>
+/// A notification that is sent when specific workflow definition versions are about to be updated.
+/// </summary>
+[PublicAPI]
+public record WorkflowDefinitionVersionsUpdating(IEnumerable<WorkflowDefinitionVersionsUpdate> VersionUpdate) : INotification;
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Notifications/WorkflowDefinitionVersionsUpdating.cs:10]: `VersionUpdate` should be renamed to `VersionUpdates` (plural).

File: src/modules/Elsa.Workflows.Management/Contracts/IActivityRegistryPopulator.cs (hunk 13 above)
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Management/Contracts/IActivityRegistryPopulator.cs:39]: Additionally, the "populater" exists only to populate the registry, so either we need to rename the service to e.g. "updater", or introduce a separate service called e.g. "depopulator" or "trimmer".

File: src/modules/Elsa.Workflows.Management/Services/WorkflowDefinitionManager.cs [hunk 17]
```diff
@@ -45,7 +45,6 @@ public async Task<long> DeleteByDefinitionIdAsync(string definitionId, Cancellat
         await _notificationSender.SendAsync(new WorkflowDefinitionDeleting(definitionId), cancellationToken);
         var filter = new WorkflowDefinitionFilter { DefinitionId = definitionId };
         var count = await _store.DeleteAsync(filter, cancellationToken);
-        await EnsureLastVersionIsLatestAsync(definitionId, cancellationToken);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Management/Services/WorkflowDefinitionManager.cs:48]: Good thinking 👍🏻

**Review by sfmskywalker — CHANGES_REQUESTED** [review]

File: src/modules/Elsa.Workflows.Core/Contracts/IActivityRegistryUpdateService.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+/// </summary>
+public interface IActivityRegistryUpdateService
+{
+    /// <summary>
+    /// Tries to add a workflow as an activity to the registry.
+    /// </summary>
+    /// <param name="providerType">The type of the activity provider.</param>
+    /// <param name="workflowDefinitionId">The ID of the workflow definition.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    Task AddToRegistry(Type providerType, string workflowDefinitionId, CancellationToken cancellationToken = default);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Contracts/IActivityRegistryUpdateService.cs:14]: Given that this service is specifically designed to update the Activity Registry with Activity Descriptors provided by the WorkflowDefinitionActivityProvider, we can remove the providerType parameter and instead hardcode its use (`typeof(WorkflowDefinitionActivityProvider)`) in the implementing class.

On that note, we should probably also rename this service (IActivityRegistryUpdateService) to something like `WorkflowDefinitionActivityRegistryUpdater`, given that it's not a general-purpose activity registry updater, but rather, a service that handles workflow definition changes by surgically updating the activity registry.

File: src/modules/Elsa.Workflows.Core/Services/ActivityRegistryUpdateService.cs [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        var provider = providers.First(x => x.GetType() == providerType);
+        var descriptors = await provider.GetDescriptorsAsync(cancellationToken);
+        var descriptorToAdd = descriptors
+            .SingleOrDefault(d =>
+                d.CustomProperties.TryGetValue("WorkflowDefinitionVersionId", out var val) &&
+                val.ToString() == workflowDefinitionVersionId);
+        
+        if (descriptorToAdd is not null)
+            registry.Add(providerType, descriptorToAdd);
```
Comment (sfmskywalker) [inline src/modules/Elsa.Workflows.Core/Services/ActivityRegistryUpdateService.cs:22]: Keeping my previous comment in mind, this line would become `registry.Add(typeof(WorkflowDefinitionActivityProvider), descriptorToAdd)`.

**Review by sfmskywalker — CHANGES_REQUESTED** [review]
In general, looking good. Just a small request that'll make it tight.

**Pull request merged**